api/apis/integration/route_test.go

```diff
@@ -327,4 +301,118 @@ var _ = Describe("Route Handler", func() {
 			})
 		})
 	})
+
+	Describe("POST /v3/routes/{guid}/destinations endpoint", func() {
+
+		var (
+			domainGUID string
+			routeGUID1 string
+			cfRoute1   *networkingv1alpha1.CFRoute
+
+			newDestinationAppGUID string
+		)
+
+		BeforeEach(func() {
+
+			domainGUID = generateGUID()
+			routeGUID1 = generateGUID()
+
+			cfDomain := &networkingv1alpha1.CFDomain{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      domainGUID,
+					Namespace: rootNamespace,
+				},
+				Spec: networkingv1alpha1.CFDomainSpec{
+					Name: "foo.domain",
+				},
+			}
+			Expect(
+				k8sClient.Create(ctx, cfDomain),
+			).To(Succeed())
+			// give test user access to the Domain
+			createRoleBinding(ctx, userName, spaceDeveloperRole.Name, rootNamespace)
+
+			cfRoute1 = &networkingv1alpha1.CFRoute{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      routeGUID1,
+					Namespace: namespace.Name,
+				},
+				Spec: networkingv1alpha1.CFRouteSpec{
+					Host:     "my-subdomain-1",
+					Path:     "",
+					Protocol: "http",
+					DomainRef: corev1.ObjectReference{
+						Name:      domainGUID,
+						Namespace: rootNamespace,
+					},
+					Destinations: []networkingv1alpha1.Destination{
+						{
+							GUID: "destination-guid",
+							Port: 8080,
+							AppRef: corev1.LocalObjectReference{
+								Name: "some-app-guid",
+							},
+							ProcessType: "web",
+							Protocol:    "http1",
+						},
+					},
+				},
+			}
+			Expect(
+				k8sClient.Create(ctx, cfRoute1),
+			).To(Succeed())
+
+			newDestinationAppGUID = "new-destination-app-guid"
+			requestBody := fmt.Sprintf(`{
+						"destinations": [
+							{
+								"app": {
+									"guid": %q
+								},
+								"protocol": "http1"
+							}
+						]
+					}`, newDestinationAppGUID)
+
+			var err error
+			req, err = http.NewRequestWithContext(ctx, "POST", serverURI("/v3/routes/"+routeGUID1+"/destinations"), strings.NewReader(requestBody))
+			Expect(err).NotTo(HaveOccurred())
+			req.Header.Add("Content-type", "application/json")
+		})
+
+		JustBeforeEach(func() {
+			router.ServeHTTP(rr, req)
+		})
+
+		When("the user is not authorized in the space", func() {
+			It("returns a not found status", func() {
+				Expect(rr).To(HaveHTTPStatus(http.StatusNotFound))
+				Expect(rr).To(HaveHTTPBody(ContainSubstring("Route not found")), rr.Body.String())
+			})
+		})
+
+		When("the user has readonly access to the route", func() {
+			BeforeEach(func() {
+				createRoleBinding(ctx, userName, spaceManagerRole.Name, namespace.Name)
+			})
+
+			It("returns a forbidden error", func() {
+				Expect(rr).To(HaveHTTPStatus(http.StatusForbidden))
+			})
+		})
+
+		When("the user is a space developer", func() {
```

Comment:
I am not super enthusiastic about the approach where we test the happy path only in e2e and then test only failure cases in the integration tests. I don't feel like this represents a meaningful de-duplication, because the majority of code & effort is spent in test setup, so just removing the happy path test doesn't buy much. I also feel like without the happy path test in the integration suite, we don't have assurance that we've written the test setup correctly, so even though the failure case fails, we don't know for sure that it is failing for the reason we intended.

I think we are planning some meetings/discussions about overall testing strategies. I wonder if more changes here might wait until we have agreed on an overall approach for the e2e versus integration tests?

Reply:
Yep - fine with leaving the happy path in the integration test.

I'm more concerned that we don't have the e2e test. That would give us confidence that the cluster role yamls are correct, and the components behave as expected when using the wiring from main.go.